File: fixtures/sample-repos/cypress/cypress/pages/loginActions.js
import { expect } from 'chai';
import { loginLocators } from './loginLocators';

// Service object: receives the locator helpers, owns user workflows.
// No assertions here — specs own `.should()`.
export class LoginActions {
  visit(path = '/login') {
    cy.visit(path);
  }

  signIn(email, password) {
    loginLocators.emailInput().clear().type(email);
    loginLocators.passwordInput().clear().type(password);
    loginLocators.signInButton().click();
  }
}

File: fixtures/sample-repos/appium-dotnet/Pages/Actions/LoginActions.cs
using OpenQA.Selenium.Appium;
using Sample.AppiumDotnet.Pages.Locators;

namespace Sample.AppiumDotnet.Pages.Actions;

public class LoginActions
{
    private readonly LoginLocators _locators;

    public LoginActions(AppiumDriver driver)
    {
        _locators = LoginLocators.For(driver);
    }

    public static LoginActions For(AppiumDriver driver) => new(driver);

    public void SignIn(string email, string password)
    {
        _locators.EmailInput.SendKeys(email);
        _locators.PasswordInput.SendKeys(password);
        _locators.SignInButton.Click();
    }
}
